packages/api-bindings/src/apollo/cache/utils/cursorBasedPagination.ts

```diff
@@ -27,34 +23,71 @@ export function cursorBasedPagination<T extends CursorBasedPagination>(
 
       // items that are not in the cache anymore (for .e.g deleted publication)
       const danglingItems = items.filter((item) => !canRead(item));
-      const readRes = {
+      const readRes: SafeReadonly<TResult> = {
         ...existing,
         items,
         pageInfo: {
           ...pageInfo,
+          beforeCount: existing.pageInfo.beforeCount ?? 0,
+          moreAfter: existing.pageInfo.moreAfter ?? existing.pageInfo.next !== null,
           // reduce total count by excluding dangling items so it won't cause a new page query
           // after item was removed from the cache (for .e.g deleted publication)
           totalCount:
             pageInfo.totalCount !== null ? pageInfo.totalCount - danglingItems.length : null,
         },
-      } as SafeReadonly<T>;
+      };
 
       return readRes;
     },
 
-    merge(existing: Readonly<T> | undefined, incoming: SafeReadonly<T>) {
-      if (!existing) {
-        return incoming;
+    merge(
+      existing: Readonly<TResult> | undefined,
+      incoming: SafeReadonly<TResult>,
+      { variables = {} }: FieldFunctionOptions,
+    ) {
+      if (!isCursor(variables.cursor) || !existing) {
+        return {
+          ...incoming,
+          pageInfo: {
+            ...incoming.pageInfo, // future-proofing in case we add more fields to pageInfo
+            beforeCount: 0, // assume there is no newer results than the provided
+            moreAfter: incoming.pageInfo.next !== null,
+          },
+        };
       }
 
       const existingItems = existing.items;
       const incomingItems = incoming.items;
 
-      return {
-        ...incoming,
-        items: existingItems.concat(incomingItems),
-        pageInfo: incoming.pageInfo,
-      } as SafeReadonly<T>;
+      if (variables.cursor === existing.pageInfo.prev) {
+        return {
+          ...incoming,
+          items: incomingItems.concat(existingItems),
+          pageInfo: {
+            ...incoming.pageInfo, // future-proofing in case we add more fields to pageInfo
+            beforeCount: incoming.pageInfo.beforeCount,
+            moreAfter: existing.pageInfo.next !== null,
+            next: existing.pageInfo.next,
+            prev: incoming.pageInfo.prev ?? existing.pageInfo.prev,
```

Comment:
Could it end up with duplicated items in the list? 
Let's assume the following flow
1. `prev` is called, 2 items returned and appended, `incoming.pageInfo.prev` is undefined so we reuse existing prev
2. X minutes later we have 2 new publications 
3. `prev` is called, 4 items returned (two we already had, two new), now you end up with duplicated items in the `items` list.

Is that a possible scenario? Maybe I'm missing some part that's doing deduplication

Reply:
I see the possible confusion, and at first I thought after reading your scenario I thought it was possible. Then remembered.

The thing is that if `items` is populated then `prev` and `next` are also populated. `prev` and `next` are null only when `items: []`.

I totally get this is not obvious from the code above. It relies on an implicit behaviour of the API. I think we can make it more obvious.

Reply:
got it, now i remember too that we had to call the api one last time to get empty cursor and empty array.
Would that be possible to make some branching logic that clearly indicates that behaviour?